### PR TITLE
Added redshift_error to Obj model and redshift history update function in candidate.py

### PR DESCRIPTION
**Pull request by scizen9**
Need to add redshift_error to tests and propagate into front end display.

### CODE REVIEW

**Comment by scizen9** [comment]
Here are some screenshots: 

1. Display on source page:
image

2. Updating redshift
image

3. Redshift history
image

**Comment by scizen9** [comment]
Will fix [issue #1795 ](https://github.com/skyportal/skyportal/issues/1795)

File: static/js/components/UpdateSourceRedshift.jsx [hunk 1]
```diff
@@ -33,6 +33,10 @@ const UpdateSourceRedshift = ({ source }) => {
   const [state, setState] = useState(
     source.redshift ? String(source.redshift) : ""
   );
+  const [uncstate, setUncState] = useState(
```
Comment (acrellin) [inline static/js/components/UpdateSourceRedshift.jsx:36]: Actually, probably best to combine redshift and error into a single state object. Then add relevant "name" attrs to the input fields, and update using event.target.name as the state object key (see https://github.com/skyportal/skyportal/blob/master/static/js/components/NewGroupForm.jsx#L60-L67)

**Review by acrellin — COMMENTED** [review]
Looks good overall, just one request regarding the React state structure below. Also needs a test (can just expand the existing test)

File: static/js/components/UpdateSourceRedshift.jsx (hunk 1 above)
Comment (scizen9) [reply static/js/components/UpdateSourceRedshift.jsx]: Great suggestion.  Forced me to learn more about react and update my firefox after 8 years of neglect!

**Comment by scizen9** [comment]
Still need migration script, however.

**Comment by scizen9** [comment]
Apart from errors from 'black' when generating the migration script, it seems to be ready.

**Review by acrellin — APPROVED** [review]
Looks great, thanks @scizen9 . Can you also add a test where we update redshift but leave the error blank? Can just copy existing test and removing the error-related lines

File: static/js/components/UpdateSourceRedshift.jsx [hunk 2]
```diff
@@ -113,12 +150,16 @@ const UpdateSourceRedshift = ({ source }) => {
...
+                  handleSubmit({ redshift: null, redshift_error: null });
                 }}
                 startIcon={<ClearIcon />}
                 size="large"
                 data-testid="nullifyRedshiftButton"
-                disabled={isSubmitting || source.redshift === null}
+                disabled={
+                  isSubmitting ||
+                  source.redshift === null ||
+                  source.redshift_error === null
```
Comment (guynir42) [inline static/js/components/UpdateSourceRedshift.jsx:161]: Just making sure: the button is disabled if the redshift error is null? event if the redshift itself has a value?

**Review by guynir42 — APPROVED** [review]
Very minor suggestions. Other than that it looks good to me.

File: static/js/components/UpdateSourceRedshift.jsx (hunk 2 above)
Comment (acrellin) [reply static/js/components/UpdateSourceRedshift.jsx:161]: Good catch -- we should definitely allow the error to be null

File: skyportal/handlers/api/candidate.py [hunk 3]
```diff
@@ -41,6 +41,9 @@ def update_redshift_history_if_relevant(request_data, obj, user):
                 "set_by_user_id": user.id,
                 "set_at_utc": datetime.datetime.utcnow().isoformat(),
                 "value": request_data["redshift"],
+                "uncertainty": request_data["redshift_error"]
```
[the remark this answers is not in the document]
Comment (scizen9) [reply skyportal/handlers/api/candidate.py]: Yes.  Updated.

File: static/js/components/UpdateSourceRedshift.jsx (hunk 2 above)
Comment (scizen9) [reply static/js/components/UpdateSourceRedshift.jsx:161]: This hinges on a policy decision.  In my opinion, one should not enter a redshift without an error, but others may differ on that.  I'm open to discussion on this.

Comment (acrellin) [reply static/js/components/UpdateSourceRedshift.jsx:161]: I think for now we should allow it to be nullable, and we can revisit it later if needed

**Review by acrellin — CHANGES_REQUESTED** [review]
Marking this as in discussion, per the nullable redshift conversation

File: static/js/components/UpdateSourceRedshift.jsx (hunk 2 above)
Comment (scizen9) [reply static/js/components/UpdateSourceRedshift.jsx]: Implemented null entry for redshift_error, and added test for same.

**Comment by acrellin** [comment]
I think we should only render the plus/minus when there's a redshift error. Here it's rendered w/out a redshift value or error: 
image

**Comment by acrellin** [comment]
Also, should the redshift error be factored into the DM and DL values by adding an associated error to those?

image

**Comment by scizen9** [comment]
Current behaviour:

1. No redshift:
image

2. Redshift, no error:
image

3. Redshift and error:
image

**Review by acrellin — APPROVED** [review]

**Pull request merged**